delta-app/src/main/java/io/cdap/delta/app/CapacityBoundedEventQueue.java

```diff
@@ -0,0 +1,179 @@
+   * limit. When calculating the capacity of an event only the size of the row and previous row in DML event are
+   * calculated. Because other fields are fixed size and could be limited by queue size (the number of event in the
+   * queue)
+   */
+  public class CapacityBoundedEventQueue extends ArrayBlockingQueue<Sequenced<? extends ChangeEvent>> {
+
+    private final long capacityLimit;
+    private long capacity;
+    /** Main lock guarding all access */
+    private final ReentrantLock lock;
+    /** Condition for waiting puts */
+    private final Condition notFull;
+
+    public CapacityBoundedEventQueue(int queueSize, long capcityLimit) {
+      super(queueSize);
+      this.capacityLimit = capcityLimit;
+      this.capacity = 0;
+      this.lock = new ReentrantLock();
+      this.notFull = lock.newCondition();
+    }
+
+    private static void checkNotNull(Sequenced<? extends ChangeEvent> event) {
+      if (event == null) {
+        throw new NullPointerException();
+      }
+    }
+
+    private static int computeCapacity(Sequenced<? extends ChangeEvent> event) {
+      if (event == null) {
+        return 0;
+      }
+      ChangeEvent changeEvent = event.getEvent();
+      if (changeEvent instanceof DDLEvent) {
+        return 0;
+      }
+      return ((DMLEvent) changeEvent).getSizeInBytes();
+    }
+
+    @Override
+    public void put(Sequenced<? extends ChangeEvent> event) throws InterruptedException {
+      checkNotNull(event);
+      final ReentrantLock lock = this.lock;
+      lock.lockInterruptibly();
+      try {
+        int eventCapacity = computeCapacity(event);
+        while (capacity + eventCapacity > capacityLimit) {
+          notFull.await();
+        }
+        super.put(event);
+        capacity += eventCapacity;
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    @Override
+    public boolean offer(Sequenced<? extends ChangeEvent> event) {
+      checkNotNull(event);
+      final ReentrantLock lock = this.lock;
+      lock.lock();
+      try {
+        int eventCapacity = computeCapacity(event);
+        if (capacity + eventCapacity > capacityLimit) {
+          return false;
+        } else {
+          boolean result = super.offer(event);
+          capacity += eventCapacity;
+          return result;
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    @Override
+    public boolean offer(Sequenced<? extends ChangeEvent> event, long timeout, TimeUnit unit)
+      throws InterruptedException {
+      checkNotNull(event);
+      long nanos = unit.toNanos(timeout);
+      final ReentrantLock lock = this.lock;
+      lock.lockInterruptibly();
+      try {
+        int eventCapacity = computeCapacity(event);
+        while (capacity + eventCapacity > capacityLimit) {
+          if (nanos <= 0) {
+            return false;
+          }
+          nanos = notFull.awaitNanos(nanos);
+        }
+        boolean result = super.offer(event, nanos, TimeUnit.NANOSECONDS);
```

Comment:
to honor the contract more accurately, should actually be timing the loop above and subtracting that from the timeout sent here.

Reply:
it's already subtracted. The line 
`nanos = notFull.awitNanos(nanos)`
actually update the nanos by subtracting the waiting time.

Reply:
ah, got it